core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -317,6 +333,38 @@ public UpdateSchema unionByNameWith(Schema newSchema) {
     return this;
   }
 
+  @Override
+  public UpdateSchema addIdentifierField(String name) {
+    Types.NestedField field = schema.findField(name);
+    if (field == null) {
+      field = adds.get(TABLE_ROOT_ID).stream()
+          .filter(f -> f.name().equals(name))
+          .findAny().orElse(null);
```

Comment:
This assumes that the field is not a newly added field within a struct.

I think that this should use the same strategy that we use for `addColumn`. There are two variations of `addColumn`, one that accepts a parent name and a field name, and one that accepts only a field name. If the latter is called with a name that contains `.`, then an exception is thrown because the reference is ambiguous. For example, `a.b` could be a top-level field named `a.b` or could be field `b` nested within field `a`.

By structuring the methods that way, we always know the parent field and can use that instead of assuming `TABLE_ROOT_ID` here.

Reply:
Combining with your last comment, it seems like the best way to go is to just check the column isn't nested in a map/list, and in that case the root will always be TABLE_ROOT_ID. With that check added, I feel there is a bit redundant to have 2 methods like `addColumn`, because the private `addIdentifierField` method would always use root as parent. Let me know what you think with the new code.

Reply:
A column could be nested in a struct without being in a map or a list. For example, if you had a table of HTTP logs, you might have a column of `request` information with `request.path` and `request.headers`, along with a `response` column that has `response.code`. It would be valid to refer to `response.code` because there is only one per row. That's the case where we need the parent so we can distinguish between a top-level `response.code` and `code` within the `response` struct.

Reply:
I did not realize that we could add a single column name `response.code` in this `addColumn` API (Checked the code, sounds like the correct way to add `response.code` as top-level column is: `addColumn(null, "response.code",  type, doc)` ).

If we wanna to distinguish between the nested columns  and the top-level column whose name contains DOT,  then we will need to add extra API for this , right ? 

```java
// This is used to add columns that does not contains DOT.
UpdateSchema setIdentifierFields(Colleciton<String> names);

// This is used to add nested columns or top-level columns that contains DOT
UpdateSchema setIdentifierFields(Collection<Pair<String, String>> parentAndNames); 
```

Reply:
@openinx, that's what I was thinking at first as well, but I don't think that we need a version with parent now.

The reason is that once a column is added, it is uniquely identified by a dotted name: `a.b` might be a struct named `"a` with a field named `b` or might be a field named `a.b`, but it cannot be both. We guarantee this so that we never have ambiguous names.

Because we know that the string passed in here will always identify just one field, we don't need to handle the ambiguous case by adding an explicit parent like we do with `addColumn`.

I think that means we can have a much nicer API:
* `setIdentifierFields(String...)` and `setIdentifierFields(Collection<String>)`
* `addIdentifierField(String)`